Package python code as module 'sidh'

ioerror opened the pull request:
This is now a functional python module with many automatically generated
commands as documented in the README.md file.

ioerror committed:
Package python code as module 'sidh'

This is now a functional python module with many automatically generated
commands as documented in the README.md file.

ioerror committed:
Reformat code automatically with black

black -S -l79 sidh/*.py sidh/csidh/*.py sidh/bsidh/*.py

JJChiDguez commented:
Thanks for those improvements to easy and intuitive usage of this python-code repository!
I approve the pull request, I have tested it and everything seems to correctly work.

Now, regarding the data missing (the strategies ones), the current version of the code allows to generated/precomputed strategies whenever it is a new instance.
For example, `python3 main.py -p b6 -f svelu -a bsidh` will create the strategy `bsidh-b5-svelu-classical`

Anyhow, from your modifications, it would be enough to run in root user, I mean: `sudo sidh -p b5 -f svelu -a bsidh
` will create the strategy field.

For that reason, I suggest avoiding (for future pull requests) installation using sudo. But we can improve it later :wink: 

PS. I hope by Wednesday to merge the pull request :sweat_smile:

ioerror committed:
begin major refactor

- Add CHANGELOG.md
- Add CONTRIBUTORS
- Update setup.py
- refactor into object oriented style

leif committed:
_fp.py: make OO, remove globals

this breaks the tests, of course, and is to be combined with jake's in-progress work

ioerror committed:
CSIDH test interface is now functional

sidh/csidh/__init__.py now contains doctests for a possible CSIDH API.
The test uses montgomery, hvelu, p512, df for params.

ioerror commented:
The last three commits begin a major refactor into a more object oriented style. This surfaces a CSIDH api which will allow other python programs to compute a DH with montgomery/p512/hvelu/df as parameters. Currently the api does not have a key generation or serialization format - so it is not yet a fully functional api. Additional changes to come soon.

leif committed:
add tvelu

trying this out in a functional-programming sort of style, instead of proper OO
python, so that less needs to be done.

leif committed:
add _gae_wd1

also in an FP style, for now at least. i'm not sure if I like it better with
this faux-OO style enabled by py3's nonlocal keyboard, but it sure makes things
quicker to refactor.

leif committed:
make _montgomery more OO

leif committed:
make _montgomery less OO

will decide tomorrow if we want to revert this commit... probably

going from OO to this paradigm is easy; basically it's just s/self\.//g and
s/\(self, /(/ ... but going the other way is a lot of work. so, probably we
should revert this commit for now and discuss it with JJChiDguez before
going ahead further with this unorthodox style of python.

leif committed:
remove remaining import *

_poly_{mul,redc} still have a global fp, though

ioerror commented:
The last five commits above are in an alternative functional style and it probably makes sense to pick one of the two (though they can be mixed together) for finishing the refactoring into a library.

ioerror committed:
rename sidh._math to sidh.math

leif committed:
rename {_,}*py

this breaks existing functionality, but is a step in the right direction.
probably we should squash our work since the initial move into sidh/ into a
single commit to omit the existence of the _*py files at all.

ioerror committed:
wip click rewrite

ioerror committed:
svelu genkey works

ioerror committed:
add svelu test that pass

ioerror committed:
hide future click options until they are needed

ioerror committed:
other hide future click options until they are needed

leif committed:
WIP

leif committed:
WIP bench.py

ioerror committed:
wip on bounds

ioerror committed:
wip

leif committed:
WIP

leif committed:
back to system data path

ioerror committed:
bounds runs

ioerror committed:
add csidh_header command

leif committed:
bench command does something

ioerror committed:
add click command for csidh-ijk; does not fully work

leif committed:
remove remaining globals from gae_df.py

ioerror committed:
fix missing self reference

ioerror committed:
fixes to hvelu; add not working print_parameters

leif committed:
delete old file

leif committed:
add pytest.ini and a test

pytest won't run doctests currently.
(current doctests take a long time...)

leif committed:
templated tests

most fail right now, but this shows how we can write tests once and run them for everything

leif committed:
more tests, some black formatting

leif committed:
run all tests by default

now that i know how to use pytest -k (see comments for how to)

the test_genpubvalidate fails for df and wd2 currently but passes for wd1

ioerror committed:
Proper name for Montgomery curve

leif committed:
use SystemRandom; update TODO

leif committed:
df and wd2 fixes

test_genpubvalidate passes now

ioerror committed:
port previous subcommands to new click api

ioerror committed:
remove unused docstring that cconfused click

leif committed:
enable doctests in pytest.ini

also restored framework.py for reference, but commented it all out

ioerror committed:
add final subcommand csidh-main

leif committed:
re-remove framework.py

ioerror committed:
implement top level printstrategy command for csidh

leif committed:
fix verbose/suitable mode

and add tests for it

ioerror committed:
update gitignore

leif committed:
cleanup tests

    fixed docstring
    black -S -l79 test/*py

ioerror committed:
all examples in README.md run except one

ioerror committed:
remove static keys; update TODO

ioerror committed:
update TODO

ioerror committed:
add cli api; add CSIDH library api

ioerror committed:
replace default random (MT rng) with a csprng

leif committed:
add test-cli.sh

leif committed:
update README.md

that isn't broken anymore

leif committed:
remove last globals

this creates faux-oo objects for the poly_mul and poly_redc modules to store
the F_p object they require. they're faux-oo instead of proper classes to avoid
needing to pollute the code with 'self.' everywhere; i still am not sure if i
like this this approach.

prior to this commit, a CSIDH object would stop working correctly after another
CSIDH object was instantiated with a different prime.

this change also might fix a bug in our previous half-refactored
poly_mul_init() code, in a codepath in qring_mul which we apparently aren't
reaching in our tests, where the 'negative_powers_of_two' symbol was previously
undefined but now is in-scope.

ioerror committed:
add multievaluation and tuned options to cli/library api

leif committed:
make F_p only know about p

leif committed:
reformat code with black

the result of running this command:
black -S -l79 sidh/csidh/*py sidh/fp.py sidh/constants.py sidh/__main__.py

leif committed:
enable no-multievaluation tests

previously they had colliding names with the multievaluation tests so they
weren't being run.

leif committed:
fix docstring

leif committed:
fix sdacs

from F_p refactor

leif committed:
remove separate validate test
